Fixed #27083 -- Added basic support for weak ETags.

marfire opened the pull request:
https://code.djangoproject.com/ticket/27083

timgraham commented:
buildbot, add to whitelist.

charettes commented:
The tests are failing on Python 2.7 because of your usage of `fullmatch`.

marfire commented:
@charettes: Thanks, I've updated the PR and started testing with 2.7 locally.

I also added additional tests and documentation. Still missing is an addition to the release notes.

marfire commented:
By the way, apologies if I've been spamming the build server. I was using my Github fork to pass work between two servers (dev and test) without realizing that each push updated the PR.

timgraham commented:
If it's possible to break this up into some smaller commits, that could make review much easier, especially for a non-ETag expert like me. :-)

marfire commented:
Ha, this *was* my attempt to keep things small! :-O (As I'm also planning to address the ETag issues discussed [here](https://code.djangoproject.com/ticket/19705#comment:14)...)

I think this could still reasonably be divided into two parts: one adding support for the use of weak ETags, and the other updating the `get_conditional_response()` algorithm to match RFC 7232. Does that sound sensible?

The reason I did both at the same time is that adding weak ETag support required changing `get_conditional_response()` anyway, so I thought I'd go ahead and change it fully rather than update the old version just to throw it away with the next commit. But I don't mind splitting it if you'd find the review easier...

timgraham commented:
Whatever you think, I don't want to waste your time.

marfire commented:
OK, I split the commit. This one adds syntactic support for weak ETags but doesn't change the complicated comparison algorithm.

It touches a fair number of lines, but many of the changes are trivial (e.g. removing quotes from comparisons in the tests). Hopefully it will be easier to review...

timgraham commented:
Is it ready for review then? (Not sure since the ticket flags aren't updated.)

marfire commented:
Good point, I neglected to do that. It's ready, and I will update the flags.

As I mentioned above, I'm not sure what merits a mention in the release notes. There is technically a backwards incompatibility here (see the last paragraph of the ticket description). Should I put a mention under Backwards incompatible changes in 1.11 -> Miscellaneous?

timgraham commented:
Sure, unless there's more a sentence of two of explanation in which case you could create your own heading.

timgraham commented on django/utils/http.py:
```diff
@@ -19,7 +19,7 @@
 )
 
-ETAG_MATCH = re.compile(r'(?:W/)?"((?:\\.|[^"])*)"')
+ETAG_MATCH = re.compile(r'\A((?:W/)?"[^"]*")\Z')  # based on RFC 7232, Appendix C
```
Could you annotate the regex similar to:
```
tld_re = (
    '\.'                                # dot
    '(?!-)'                             # can't start with a dash
    '(?:[a-z' + ul + '-]{2,63}'         # domain label
    '|xn--[a-z0-9]{1,59})'              # or punycode label
    '(?<!-)'                            # can't end with a dash
    '\.?'                               # may have a trailing dot
)
```

timgraham commented on django/utils/http.py:
```diff
@@ -90,7 +90,7 @@ def urlencode(query, doseq=0):
     return original_urlencode(
         [(force_str(k),
-         [force_str(i) for i in v] if isinstance(v, (list, tuple)) else force_str(v))
+          [force_str(i) for i in v] if isinstance(v, (list, tuple)) else force_str(v))
```
It would be better not to make whitespace changes like this.

timgraham commented on django/utils/http.py:
```diff
@@ -232,30 +232,31 @@ def urlsafe_base64_decode(s):
-    Parses a string with one or several etags passed in If-None-Match and
-    If-Match headers by the rules in RFC 2616. Returns a list of etags
-    without surrounding double quotes (") and unescaped from \<CHAR>.
+    Parses a string of ETags given in an If-None-Match or If-Match header as
```
When changing docstrings, we are updating them to use PEP 257 verb style: Parse / Return, etc.

timgraham commented on django/utils/http.py:
```diff
@@ -232,30 +232,31 @@ def urlsafe_base64_decode(s):
+        # This is just enough validation to prevent outdated RFC 2616 ETags
+        # from breaking things.
+        etag_matches = (ETAG_MATCH.match(etag.strip()) for etag in etag_str.split(","))
+        etags = [match.group(1) for match in etag_matches if match]
```
return without intermediate `etags` variable

timgraham commented on django/utils/http.py:
```diff
@@ -232,30 +232,31 @@ def urlsafe_base64_decode(s):
+    if etag_str.strip() == '*':
+        return ['*']
+    else:
+        # This is just enough validation to prevent outdated RFC 2616 ETags
```
prevent ETags in the old RFC 2616 format

timgraham commented on django/utils/http.py:
```diff
@@ -232,30 +232,31 @@ def urlsafe_base64_decode(s):
+        return ['*']
+    else:
+        # This is just enough validation to prevent outdated RFC 2616 ETags
+        # from breaking things.
```
Perhaps "breaking things" could be better described. Is the idea that they should still be matched?

marfire replied:
I'll reword this. What I'm getting at is that splitting on `,` allows each tag to be matched or not individually, whereas doing a findall() like before allows one weird RFC 2616 tag to break all the others (e.g. `"\"", "X"` would become `"\"` and `", "`).

timgraham commented on django/utils/http.py:
```diff
@@ -232,30 +232,31 @@ def urlsafe_base64_decode(s):
+    else:
+        # This is just enough validation to prevent outdated RFC 2616 ETags
+        # from breaking things.
+        etag_matches = (ETAG_MATCH.match(etag.strip()) for etag in etag_str.split(","))
```
use single quotes for consistency

timgraham commented on django/utils/http.py:
```diff
@@ -232,30 +232,31 @@ def urlsafe_base64_decode(s):
+    if ETAG_MATCH.match(etag_str):
+        return etag_str
+    else:
+        # We don't validate etag_str.
```
I'm trying to avoid "We" in comments. I think "Don't validate etag_str." has the same meaning.

timgraham commented on django/views/decorators/http.py:
```diff
@@ -58,20 +58,20 @@ def condition(etag_func=None, last_modified_func=None):
     The parameters are callables to compute the ETag and last modified time for
     the requested resource, respectively. The callables are passed the same
-    parameters as the view itself. The Etag function should return a string (or
+    parameters as the view itself. The ETag function should return a string (or
```
It would be appropriate to make the renames in a separate commit unless this is the only one to be fixed in which case it's probably not worth separating it.

timgraham commented on django/middleware/http.py:
```diff
@@ -11,6 +11,7 @@ class ConditionalGetMiddleware(MiddlewareMixin):
 
     Also sets the Date and Content-Length response-headers.
     """
+
```
Please revert whitespace change.

timgraham replied:
The guideline is "don't include unrelated cosmetic changes in bug/feature patch."

timgraham commented on docs/releases/1.11.txt:
```diff
@@ -483,6 +483,10 @@ Miscellaneous
   ``migrate`` command. Use the new :djadmin:`remove_stale_contenttypes` command
   instead.
 
+* ETags are now parsed according to the updated RFC 7232 Conditional Requests
```
RFC 7232 ->
```
:rfc:`7232`
```

timgraham commented on docs/releases/1.11.txt:
```diff
@@ -483,6 +483,10 @@ Miscellaneous
 
+* ETags are now parsed according to the updated RFC 7232 Conditional Requests
+  specification. The ETag syntax there is slightly different from that of the
+  older RFC 2616 specification which was used in previous versions.
```
Is it worth giving an example of how this might affect a project or will it be obvious to ETag users?

marfire replied:
In my opinion an example would be overkill. My guess is that this change won't affect anyone, but if it does it will be because they're supplying their own, unusually-formatted ETags, and in that case they should really be consulting the RFCs directly. Let me know if you disagree.

timgraham commented on docs/topics/conditional-view-processing.txt:
```diff
@@ -64,7 +64,15 @@ order, as the view function they are helping to wrap. The function passed
+
+.. versionchanged:: 1.11
+
+    In prior versions the return value from ``etag_func()`` was interpreted as
```
In older versions,

timgraham commented on docs/topics/conditional-view-processing.txt:
```diff
@@ -64,7 +64,15 @@ order, as the view function they are helping to wrap. The function passed
+    the unquoted part of the ETag. That prevented the use of weak ETags, which
+    have the format ``W/"<string>"``. The return value is now expected to be
+    an ETag as defined by the specification (including the quotes), although
+    the unquoted format is also accepted for backwards compatibility reasons.
```
Is it fine to remove this entire note when removing or should anything remain? If the latter, please make the note self-contained as described in https://docs.djangoproject.com/en/dev/internals/contributing/writing-documentation/#documenting-new-features.

marfire replied:
To me, the existing phrase "a string representing the ETag" implies the full quoted form as defined by the specification (including weak ETags). So going forward, I think supporting the unquoted version can be an undocumented nod to backwards compatibility, making it fine for this note to go away. Another option would be to spell out the quoted / unquoted thing explicitly in the documentation. Yet another option would be to actively deprecate support for the unquoted form (though it doesn't seem worth the trouble to me).

Let me know your opinion.

timgraham replied:
I trust your judgment.

timgraham commented on tests/conditional_processing/tests.py:
```diff
@@ -24,12 +25,16 @@ def assertFullResponse(self, response, check_last_modified=True, check_etag=True
         self.assertEqual(response.status_code, 304)
         self.assertEqual(response.content, b'')
 
+    def assertPreconditionFailed(self, response):
```
Perhaps this helper could be added in a separate commit so there are less unrelated test changes in this commit (unless it really is related).

timgraham commented on tests/conditional_processing/tests.py:
```diff
@@ -188,7 +193,32 @@ def test_single_condition_head(self):
         response = self.client.head('/condition/')
         self.assertNotModified(response)
 
+    # Test that the same ETag results whether returning quoted
```
Should these comments be docstrings? We're in the habit of stating the expected behavior rather than including "Test that..." since all tests test things.

timgraham commented on tests/utils_tests/test_http.py:
```diff
@@ -196,14 +196,22 @@ def test_is_same_domain_bad(self):
+        etags = http.parse_etags('*')
+        self.assertEqual(etags, ['*'])
+
+        # We ignore RFC 2616 tags that are now invalid.
```
Chop "We"
"now invalid" ->  "invalid according to RFC 7232"

timgraham commented on tests/conditional_processing/views.py:
```diff
@@ -27,3 +27,23 @@ def etag_view1(request):
 etag_view2 = etag(lambda r: ETAG)(etag_view2)
+
+
+# etag_func() returns an ETag without quotes
```
docstrings?

marfire commented:
Thanks for the review, Tim, I've updated the PR.

timgraham commented on tests/conditional_processing/tests.py:
```diff
@@ -188,7 +189,34 @@ def test_single_condition_head(self):
+        response_unquoted = self.client.get('/condition/unquoted_etag/')
+        self.assertEqual(response_quoted['ETag'], response_unquoted['ETag'])
+
+    # Since it's possible that the matching algorithm could use the wrong value
```
Docstring?

marfire replied:
I don't think this makes sense as a docstring. It's not explaining what the function does (or tests), it's more of a meta-comment explaining why this test exists at all (as opposed to having no additional tests and relying on `test_unquoted()`, or duplicating all of the tests for both quoted and unquoted values).

timgraham commented on tests/conditional_processing/urls.py:
```diff
@@ -8,4 +8,7 @@
     url('^condition/etag/$', views.etag_view1),
     url('^condition/etag2/$', views.etag_view2),
+    url('^condition/unquoted_etag/$', views.etag_view_unquoted),
+    url('^condition/weak_etag/$', views.etag_view_weak),
```
URL seems to be unused

marfire replied:
You're right, it's leftover from the original commit...

timgraham commented on docs/topics/conditional-view-processing.txt:
```diff
@@ -64,7 +64,15 @@ order, as the view function they are helping to wrap. The function passed
+
+.. versionchanged:: 1.11
+
+    In older versions the return value from ``etag_func()`` was interpreted as
```
versions,

timgraham commented on docs/topics/conditional-view-processing.txt:
```diff
@@ -64,7 +64,15 @@ order, as the view function they are helping to wrap. The function passed
+    the unquoted part of the ETag. That prevented the use of weak ETags, which
+    have the format ``W/"<string>"``. The return value is now expected to be
+    an ETag as defined by the specification (including the quotes), although
+    the unquoted format is also accepted for backwards compatibility reasons.
```
chop "reasons"

timgraham commented on docs/releases/1.11.txt:
```diff
@@ -488,6 +488,10 @@ Miscellaneous
   ``migrate`` command. Use the new :djadmin:`remove_stale_contenttypes` command
   instead.
 
+* ETags are now parsed according to the updated :rfc:`7232` Conditional Requests
```
Perhaps a bit simpler:
```
* ETags are now parsed according to the :rfc:`7232` Conditional Requests
  specification rather than the syntax from :rfc:`2616`.
```

marfire commented:
I've updated the PR.

By the way, instead of amending the commit, would it be better if I rebased the commit and then made the post-review changes as a separate commit? I was trying to save you the trouble of squashing multiple commits in the end, but maybe it makes review easier...

timgraham commented:
merged in 4ef0e019b7dd3d2bf93b5c705b3b7df9cdb77561, thanks!

For larger PRs, incremental commits can be helpful, but this is small enough that the choice you made is good.
